Source/WebCore/Modules/speech/SpeechSynthesis.cpp

```diff
@@ -178,16 +178,28 @@ void SpeechSynthesis::cancel()
     // Remove all the items from the utterance queue.
     // Hold on to the current utterance so the platform synthesizer can have a chance to clean up.
     RefPtr current = protectedCurrentSpeechUtterance();
-    m_utteranceQueue.clear();
     if (RefPtr speechSynthesisClient = m_speechSynthesisClient.get()) {
+        m_utteranceQueue.clear();
         speechSynthesisClient->cancel();
         // If we wait for cancel to callback speakingErrorOccurred, then m_currentSpeechUtterance will be null
         // and the event won't be processed. Instead we process the error immediately.
         speakingErrorOccurred();
         m_currentSpeechUtterance = nullptr;
-    } else if (m_platformSpeechSynthesizer)
+    } else if (m_platformSpeechSynthesizer) {
+        // Clear m_utteranceQueue before calling cancel to avoid picking up new utterances
+        // on platform completion callback
+        Deque<Ref<SpeechSynthesisUtterance>> utteranceQueue = WTFMove(m_utteranceQueue);
         m_platformSpeechSynthesizer->cancel();
 
+        // Trigger canceled events for queued utterances
+        while (!utteranceQueue.isEmpty()) {
+            Ref<SpeechSynthesisUtterance> utterance = utteranceQueue.takeFirst();
+            // Current utterance is handled in platform cancel()
+            if (current.get() != utterance.ptr())
+                utterance.get().errorEventOccurred(eventNames().errorEvent, SpeechSynthesisErrorCode::Canceled);
+        }
```

Comment:
I think we need to fire event for pending utterances in queue in all cases, no matter the synthesizer is in-process or out-of-process; i.e. this part should be moved out of the `else if m_platformSpeechSynthesizer` condition.